libvips/foreign/jxlsave.c

```diff
@@ -389,93 +754,120 @@ vips_foreign_save_jxl_sink_disc(VipsRegion *region, VipsRect *area, void *a)
 			if (vips_foreign_save_jxl_add_frame(jxl))
 				return -1;
 
-			jxl->write_y = 0;
+			jxl->scanline_y = 0;
 		}
 	}
 
 	return 0;
 }
 
 static int
-vips_foreign_save_jxl_add_metadata(VipsForeignSaveJxl *jxl, VipsImage *in)
+vips_foreign_save_jxl_save_sequential(VipsForeignSaveJxl *jxl, VipsImage *in)
 {
-#ifdef HAVE_LIBJXL_0_7
-	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS(jxl);
-	int i;
+	/* Write the header.
+	 */
+	if (vips_foreign_save_jxl_process_output(jxl))
+		return -1;
 
-	for (i = 0; i < VIPS_NUMBER(libjxl_metadata); i++)
-		if (vips_image_get_typeof(in, libjxl_metadata[i].name)) {
-			uint8_t *data;
-			size_t length;
+	/* RGB(A) frame as a contiguous buffer.
+	 */
+	jxl->scanline_size = VIPS_IMAGE_SIZEOF_LINE(in) * jxl->page_height;
+	if (!(jxl->scanline_buffer = vips_tracked_malloc(jxl->scanline_size)))
+		return -1;
 
-#ifdef DEBUG
-			printf("attaching %s ..\n", libjxl_metadata[i].name);
-#endif /*DEBUG*/
+	if (vips_sink_disc(in, vips_foreign_save_jxl_sink_disc, jxl))
+		return -1;
 
-			if (vips_image_get_blob(in,
-					libjxl_metadata[i].name, (const void **) &data, &length))
-				return -1;
+	/* This function must be called after the final frame and/or box,
+	 * otherwise the codestream will not be encoded correctly.
+	 */
+	JxlEncoderCloseInput(jxl->encoder);
 
-			/* It's safe to call JxlEncoderUseBoxes multiple times
-			 */
-			if (JxlEncoderUseBoxes(jxl->encoder) != JXL_ENC_SUCCESS) {
-				vips_foreign_save_jxl_error(jxl, "JxlEncoderUseBoxes");
-				return -1;
-			}
+	/* Flush any remaining libjxl writes.
+	 */
+	if (vips_foreign_save_jxl_process_output(jxl))
+		return -1;
 
-			/* JPEG XL stores EXIF data without leading "Exif\0\0" with offset
-			 */
-			if (!strcmp(libjxl_metadata[i].name, VIPS_META_EXIF_NAME)) {
-				if (length >= 6 && vips_isprefix("Exif", (char *) data)) {
-					data = data + 6;
-					length -= 6;
-				}
+	return 0;
+}
 
-				size_t exif_size = length + 4;
-				uint8_t *exif_data = g_malloc0(exif_size);
+static int
+vips_foreign_save_jxl_save_chunked_page(VipsForeignSaveJxl *jxl,
+	int n, VipsImage *page)
+{
+	jxl->page = page;
+	jxl->tile_hash = g_hash_table_new_full(g_direct_hash, g_direct_equal,
+			NULL, (GDestroyNotify) g_object_unref);
 
-				if (!exif_data) {
-					vips_error(class->nickname, "%s", _("out of memory"));
-					return -1;
-				}
+	JxlEncoderFrameSettings *frame_settings;
+	frame_settings = JxlEncoderFrameSettingsCreate(jxl->encoder, NULL);
+	JxlEncoderFrameSettingsSetOption(frame_settings,
+		JXL_ENC_FRAME_SETTING_DECODING_SPEED, jxl->tier);
+	JxlEncoderSetFrameDistance(frame_settings, jxl->distance);
+	JxlEncoderFrameSettingsSetOption(frame_settings,
+		JXL_ENC_FRAME_SETTING_EFFORT, jxl->effort);
+	JxlEncoderSetFrameLossless(frame_settings, jxl->lossless);
 
-				/* The first 4 bytes is offset which is 0 in this case
-				 */
-				memcpy(exif_data + 4, data, length);
+	if (jxl->info.have_animation) {
+		JxlFrameHeader header = { 0 };
```

Comment:
Nice! It looks like the `memset()` in `vips_foreign_save_jxl_add_frame()` can also be replaced with zero initialization  now (`= { 0 }`).

Reply:
ooo yes, nice spot